integration/test_web.py

```diff
@@ -0,0 +1,496 @@
+import sys
+import time
+import shutil
+import json
+import urllib2
+from os import mkdir, unlink, utime
+from os.path import join, exists, getmtime
+
+import allmydata.uri
+
+import util
+
+import requests
+import pytest_twisted
+import html5lib
+from bs4 import BeautifulSoup
+
+
+def test_index(alice):
+    """
+    we can download the index file
+    """
+    util.web_get(alice, u"")
+    # ...and json mode is valid json
+    json.loads(util.web_get(alice, u"", params={u"t": u"json"}))
+
+
+def test_upload_download(alice):
+    """
+    upload a file, then download it via readcap
+    """
+
+    FILE_CONTENTS = u"some contents"
+
+    readcap = util.web_post(
+        alice, u"uri",
+        data={
+            u"t": u"upload",
+            u"format": u"mdmf",
+        },
+        files={
+            u"file": FILE_CONTENTS,
+        },
+    )
+    readcap = readcap.strip()
+
+    data = util.web_get(
+        alice, u"uri",
+        params={
+            u"uri": readcap,
+            u"filename": u"boom",
+        }
+    )
+    assert data == FILE_CONTENTS
+
+
+def test_put(alice):
+    """
+    use PUT to create a file
+    """
+
+    FILE_CONTENTS = u"added via PUT"
+
+    resp = requests.put(
+        util.node_url(alice.node_dir, u"uri"),
+        files={
+            u"file": FILE_CONTENTS,
+        },
+    )
+    cap = allmydata.uri.from_string(resp.text.strip().encode('ascii'))
+    assert isinstance(cap, allmydata.uri.CHKFileURI)
+    assert cap.size == 153
+    assert cap.total_shares == 4
+    assert cap.needed_shares == 2
+
+
+def test_helper_status(storage_nodes):
+    """
+    successfully GET the /helper_status page
+    """
+
+    url = util.node_url(storage_nodes[0].node_dir, "helper_status")
+    resp = requests.get(url)
+    assert resp.status_code >= 200 and resp.status_code < 300
+    # XXX put in util.parse_html(resp) ...?
+    dom = BeautifulSoup(resp.content, "html5lib")
+    assert unicode(dom.h1.string) == u"Helper Status"
+
+
+def test_deep_stats(alice):
+    """
+    create a directory, do deep-stats on it and prove the /operations/
+    URIs work
+    """
+    resp = requests.post(
+        util.node_url(alice.node_dir, "uri"),
+        params={
+            "format": "sdmf",
+            "t": "mkdir",
+            "redirect_to_result": "true",
+        },
+    )
+    assert resp.status_code >= 200 and resp.status_code < 300
+
+    # when creating a directory, we'll be re-directed to a URL
+    # containing our writecap.. (XXX doesn't this violate the "URLs
+    # leak" maxim?)
```

Comment:
This XXX looks like it should be turned into a ticket or something.

Reply:
I deleted the comment .. and probably I'm just confused about this in regards to the WebAPI because other things put capabilities in the URIs too (e.g. `/uri/<cap>` appears when you upload or create anything).